### PR TITLE
Merge master into features/recursive-patterns

### PR DESCRIPTION
**Pull request by gafter**

### CODE REVIEW

**Comment by Neme12** [comment]
Oops, it looks like all the failures are tests from my PR :-/ I'll try take a look at this, but I'm not sure if I can help much because I know nothing about what's in this branch and what kind of syntax was added.

**Comment by Neme12** [comment]
One problem is this:
```c#
switch (1)
{
    case (1 + 1) w$$
}
```
this is now parsed as a DeconstructionPattern rather than an expression, which prevents `when` from being offered. I'm not sure what to do here :/

**Comment by Neme12** [comment]
The second failure looks like an actual problem with the compiler. This code doesn't seem to compile even thought it should:
```c#
class C
{
    void M()
    {
        const int var = 0;

        switch (1)
        {
            case var when true:
                break;
        }
    }
}
```

@gafter ?

**Comment by Neme12** [comment]
Actually I don't even have to involve `when` here. This code fails to compile as well:
```c#
class C
{
    void M()
    {
        const int var = 0;

        switch (1)
        {
            case var:
                break;
        }
    }
}
```

**Comment by jcouv** [comment]
I think it's fine to mark the few falling IDE tests as Skip and appropriate PROTOTYPE markers to b remind us to fix before meeting the feature.

**Comment by Neme12** [comment]
I agree we can skip the tests with a parenthesized expression for now... the whole logic for IsTypeName should probably later be updated to account for deconstructions anyway. But it looks like the 'var' tests revealed a real bug. Or this a known issue?

**Comment by gafter** [comment]
There are a few changes:
1. You cannot use a constant or type named `var` in pattern-matching.
2. You (mostly) cannot use a type named `_` in pattern-matching (only a warning in `is`).
3. You cannot declare a pattern variable named `when`.
4. We added complex recursive patterns to handle `Deconstruct` and tuples, for example `case (2) w:` matches a `Deconstruct` with a single out parameter that produces the value 2, and places the input in the variable `w`.

There are more changes, but the bottom line is that there will be a significant work item bringing your changes up-to-date with recursive pattern-matching. I'll skip these tests and add a new issue to address the work.

**Comment by Neme12** [comment]
@gafter
> You cannot use a constant or type named var in pattern-matching.

In my last example, there is no pattern matching involved:
```c#
const int var = 0;

switch (1)
{
    case var:
        break;
}
```
is `var` as a constant now disallowed in a plain old switch case as well?

**Comment by gafter** [comment]
Opened https://github.com/dotnet/roslyn/issues/26000 for the issue that the compiler does not permit a constant pattern (e.g. a constant in a switch case) named `var` any longer.

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 1]
```diff
@@ -102,7 +104,8 @@ public class WhenKeywordRecommenderTests : KeywordRecommenderTests
         [InlineData("1")]
         [InlineData("1 + 1")]
         [InlineData("true ? 1 : 1")]
-        [InlineData("(1 + )")]
+        // PROTOTYPE(patterns2): https://github.com/dotnet/roslyn/issues/25998 this resembles a positional pattern that it not handled by the recommender
+        // [InlineData("(1 + )")]
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: note: you can use Skip on InlineData as well

**Comment by CyrusNajmabadi** [comment]
Skipping these tests seems fine for now.  Once appropriate decisions are made around how the language has changed, the feature (and its tests) can be updated if necessary to reflect the intended behavior.

**Comment by gafter** [comment]
Filed https://github.com/dotnet/roslyn/issues/26001 for the Ubuntu failures.

**Comment by CyrusNajmabadi** [comment]
@gafter Can you explain this part a bit:

> We added complex recursive patterns to handle Deconstruct and tuples, for example case (2) w: matches a Deconstruct with a single out parameter that produces the value 2, and places the input in the variable w.

Could you give a couple of real world examples where this would be valuable and likely to be used?  Thanks!

**Comment by gafter** [comment]
@CyrusNajmabadi 

``` c#
    case AssignmentExpressionSyntax(SimpleNameSyntax( (var nameStringLeft) identifierTokenLeft ),
                                                           SimpleNameSyntax( (var nameStringRight) identifierTokenRight ))
            where nameStringLeft == nameStringRight:
                Error(ErrorCode.ERR_SelfAssignment, identifierTokenLeft.Location);
                break;
)
```

which can be thought as shorthand for

``` c#
    case AssignmentExpressionSyntax(SimpleNameSyntax( SyntaxToken(var nameStringLeft) identifierTokenLeft ),
                                                           SimpleNameSyntax( SyntaxToken(var nameStringRight) identifierTokenRight ))
            where nameStringLeft == nameStringRight:
                Error(ErrorCode.ERR_SelfAssignment, identifierTokenLeft.Location);
                break;
)
```

**Comment by gafter** [comment]
It is possible that we will decide to forbid that syntax (a single-element deconstruct without the type); see https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-11-20.md#syntactic-ambiguity-around-parenthesized-expression

**Comment by gafter** [comment]
This is implemented in `LanguageParser_Patterns.cs`:

``` c#
                if (type == null &&
                    propertySubpattern0 == null &&
                    designation0 == null &&
                    subPatterns.Count == 1 &&
                    subPatterns[0].NameColon == null)
                {
                    if (subPatterns[0].Pattern is ConstantPatternSyntax cp)
                    {
                        // There is an ambiguity between a deconstruction pattern `(` pattern `)`
                        // and a constant expression pattern that happens to be parenthesized.
                        // Per 2017-11-20 LDM we treat such syntax as a parenthesized expression always.
                        return _syntaxFactory.ConstantPattern(_syntaxFactory.ParenthesizedExpression(openParenToken, cp.Expression, closeParenToken));
                    }
                }

                var result = _syntaxFactory.DeconstructionPattern(type, openParenToken, subPatterns, closeParenToken, propertySubpattern0, designation0);

                // 2017-11-20 LDM decision is to disallow a deconstruction pattern that contains just a
                // single subpattern but for which the type is omitted.
                // This keeps the design space open for using parentheses for grouping patterns in the future, e.g. if we introduce `or` and
                // `and` patterns. We may add other ways to disambiguate later (e.g. a property subpattern or a trailing comma inside the parens).
                return (type == null && subPatterns.Count == 1) ? this.AddError(result, ErrorCode.ERR_SingleElementPositionalPatternRequiresType) : result;
```

**Pull request merged**